infra/terraform/main.tf

```diff
@@ -0,0 +1,321 @@
+  scope                = azurerm_container_registry.acr.id
+  role_definition_name = "AcrPull"
+  principal_id         = azurerm_user_assigned_identity.identity.principal_id
+}
+
+resource "azurerm_role_assignment" "kv_sp_secrets_user" {
+  scope                = azurerm_key_vault.kv.id
+  role_definition_name = "Key Vault Secrets User"
+  principal_id         = azurerm_user_assigned_identity.identity.principal_id
+  principal_type       = "ServicePrincipal"
+  description          = "Assign Key Vault Secrets Reader role to ACA identity"
+}
+
+resource "azurerm_container_app_environment" "aca_env" {
+  name                       = local.root_name
+  location                   = azurerm_resource_group.rg.location
+  resource_group_name        = azurerm_resource_group.rg.name
+  log_analytics_workspace_id = azurerm_log_analytics_workspace.log_analytics.id
+}
+
+resource "azurerm_container_app_environment_storage" "storage" {
+  name                         = "simulator"
+  container_app_environment_id = azurerm_container_app_environment.aca_env.id
+  account_name                 = azurerm_storage_account.storage.name
+  share_name                   = azurerm_storage_share.simulator.name
+  access_key                   = azurerm_storage_account.storage.primary_access_key
+  access_mode                  = "ReadOnly"
+}
+
+resource "azurerm_container_app" "container_app" {
+  name                         = "${var.resource_group_name}-app"
+  resource_group_name          = azurerm_resource_group.rg.name
+  container_app_environment_id = azurerm_container_app_environment.aca_env.id
+  revision_mode                = "Single"
+
+  depends_on = [
+    null_resource.docker,
+    azurerm_key_vault_secret.appinsights,
+    azurerm_key_vault_secret.openaikey,
+    azurerm_key_vault_secret.simulatorapikey
+  ]
+
+  identity {
+    type         = "UserAssigned"
+    identity_ids = [azurerm_user_assigned_identity.identity.id]
+  }
+  ingress {
+    target_port      = 8000
+    external_enabled = true
+    traffic_weight {
+      percentage      = 100
+      latest_revision = true
+    }
+  }
+  registry {
+    server   = azurerm_container_registry.acr.login_server
+    identity = azurerm_user_assigned_identity.identity.id
+  }
+
+  dynamic "secret" {
+    for_each = ["azure-openai-key", "app-insights-connection-string", "simulator-api-key"]
+    content {
+      name                = secret.value
+      key_vault_secret_id = "${azurerm_key_vault.kv.vault_uri}secrets/${secret.value}"
+      identity            = azurerm_user_assigned_identity.identity.id
+    }
+  }
+
+  secret {
+    name  = "deployment-config"
+    value = var.simulator_api_key
+  }
+
+  template {
+    min_replicas = 1
+    max_replicas = 1
+
+    container {
+      name   = "aoai-api-simulator"
+      image  = "${azurerm_container_registry.acr.login_server}/aoai-api-simulator:${var.simulator_image_tag}"
+      cpu    = 0.5
+      memory = "1Gi"
+
+      dynamic "env" {
```

Comment:
I initially wondered why you had used `dynamic` blocks here but after thinking about it more it does give a more concise result than putting an `env` block for each value!

```hcl
      env { 
        name = "AZURE_OPENAI_ENDPOINT"
        value = var.azure_openai_endpoint
      }
      env {
        name ="SIMULATOR_MODE"
        value = var.simulator_mode
      }
```